### ISSUE TITLE
Redirects drop querystring parameters 

### ISSUE THREAD

**Issue by MattSegal** [wagtail/wagtail#4414]
### Issue Summary

Wagtail's RedirectMiddleware does not preserve querystring parameters. For example:

```
requested url          www.example.com/old-path/?foo=1&bar=2
expected redirect      www.example.com/new-path/?foo=1&bar=2
actual redirect        www.example.com/new-path/
```

I am experiencing this issue on v1.13.1, but by inspecting the code it seems that it persists in v2.

### Steps to Reproduce

- Create a `Redirect` instance from `old_path` to a `redirect_page` / `redirect_link` (eg `/old-path/` --> `/new-path/`
- Browse to the old path, with query-string parameters (eg `/old-path/?foo=1&bar=2`)
- Observe that you are redirected to `/new-path/` without the original querystring parameters.

### Technical details

I believe the issue is with `RedirectMiddleware.process_response`, where the response object is passed `redirect.link`, which does not include the querystring. It should be possible to use `request.META['QUERY_STRING']`, or similar, to pass on the querystring. 

* Python version: 3.5
* Django version: 1.11.7
* Wagtail version: 1.13.1.
* Browser version: Chromium

**Comment by thibaudcolas** [comment]
Hey @MattSegal, thank you for taking the time to write such a detailed issue 🙂. It helps a lot.

I could reproduce this in Wagtail v2 – see for example: https://bakerydemo-thibaudcolas.herokuapp.com/test-redirect?test=5, which redirects to the homepage.

I see you've already attempted to look into a potential fix – would you be keen to propose a pull request?

**Comment by Vamory** [comment]
As such the fix is straight-forward, in RedirectMiddleware, replacing the following:

```
        if redirect.is_permanent:
            return http.HttpResponsePermanentRedirect(redirect.link)
        else:
            return http.HttpResponseRedirect(redirect.link)
```

with 

```
        if "?" in path:
            redirect_link = redirect.link + "?" + urlparse(path).query
        else:
            redirect_link = redirect.link

        if redirect.is_permanent:
            return http.HttpResponsePermanentRedirect(redirect_link)
        else:
            return http.HttpResponseRedirect(redirect_link)
```
works but then the following tests  breaks up:
- ERROR: test_duplicate_redirects_with_query_string_when_match_is_for_generic (wagtail.contrib.redirects.tests.TestRedirects)
- ERROR: test_duplicate_redirects_with_query_string_when_match_is_for_specific_with_qs (wagtail.contrib.redirects.tests.TestRedirects)
- ERROR: test_redirect_stripping_query_string (wagtail.contrib.redirects.tests.TestRedirects)

Basically, they've all been added to cover the cases of redirects which might use query strings for meaningful logic (redirecting from index?page_id=XXX).

Depending on the redirection policy both scenarios can be valid ones. Not sure about the best way to cover both of them though.

**Comment by thibaudcolas** [comment]
Thanks for the detective work @Vamory 👍. Looking at those tests, I found #647 which seems to cover the same thing as this issue.

I'm really no expert but ideally I would expect query strings to be preserved by default. The `?page_id=123` example is different of course. If the redirect's URL does not contain a query string, I think it would be good to preserve any incoming query string as-is.

Are there redirect implementations out there we should try to learn from?

**Comment by MattSegal** [comment]
I have a PR ready, with some examples here https://github.com/wagtail/wagtail/pull/4491

**Comment by jaroel** [comment]
This would be really useful for handling old urls using query parameters, like `/search.aspx?keywords=[word]`. ie `search.aspx?keywords=banana` -> `search/?q=banana`

**Comment by easherma-truth** [comment]
I ran into this issue too with a current project, where not keeping the querystrings is causing problems keeping analytics url params

**Comment by morezco** [comment]
I just had this issue, we wanted to use query strings to alter article behaviour.

**Comment by hunzlahmalik** [comment]
Anyone working on this? Looks simple, if not I'll like to work on it.
Recently  ran into this issue and had to override the middleware. Better if we get this option within wagtail.
